drake/solvers/NloptSolver.cpp

```diff
@@ -0,0 +1,269 @@
+                               const VariableList& variable_list) {
+  size_t var_count = 0;
+  for (const DecisionVariableView& v : variable_list) {
+    var_count += v.size();
+  }
+
+  auto tx = initializeAutoDiff(xvec);
+  TaylorVecXd this_x(var_count);
+  size_t index = 0;
+  for (const DecisionVariableView& v : variable_list) {
+      this_x.segment(index, v.size()) = tx.segment(v.index(), v.size());
+      index += v.size();
+    }
+  return this_x;
+}
+
+double EvaluateCosts(const std::vector<double>& x,
+                     std::vector<double>& grad,
+                     void* f_data) {
+  const OptimizationProblem* prog =
+      reinterpret_cast<const OptimizationProblem*>(f_data);
+
+  double cost = 0;
+  Eigen::VectorXd xvec = MakeEigenVector(x);
+
+  auto tx = initializeAutoDiff(xvec);
+  TaylorVecXd ty(1);
+  TaylorVecXd this_x(x.size());
+
+  if (!grad.empty()) {
+    grad.assign(grad.size(), 0);
+  }
+
+  for (auto const& binding : prog->getGenericObjectives()) {
+    size_t index = 0;
+    for (const DecisionVariableView& v : binding.getVariableList()) {
+      this_x.segment(index, v.size()) = tx.segment(v.index(), v.size());
+      index += v.size();
+    }
+
+    binding.getConstraint()->eval(this_x, ty);
+
+    // sam.creasey: I think the right thing to do here is to sum all
+    // of the costs if multiple objectives are specified.
+    cost += ty(0).value();
+    if (!grad.empty()) {
+      for (const DecisionVariableView& v : binding.getVariableList()) {
+        for (size_t j = v.index(); j < v.index() + v.size(); j++) {
+          // TODO sam.creasey Is summing the derivatives the right
+          // thing to do if a decision variable is bound to multiple
+          // objectives?
+          grad[j] += ty(0).derivatives()(j);
+        }
+      }
+    }
+  }
+
+  return cost;
+}
+
+/// Structure to marshall data into the NLopt callback functions,
+/// which take only a single pointer argument.
+struct WrappedConstraint {
+  const Constraint* constraint;
+  const VariableList* variable_list;
+};
+
+double EvaluateScalarConstraint(const std::vector<double>& x,
+                                std::vector<double>& grad,
+                                void* f_data) {
+  const WrappedConstraint* wrapped =
+      reinterpret_cast<WrappedConstraint*>(f_data);
+
+  Eigen::VectorXd xvec = MakeEigenVector(x);
+
+  const Constraint* c = wrapped->constraint;
+  const size_t num_constraints = c->getNumConstraints();
+  assert(num_constraints == 1);
+
+  TaylorVecXd ty(1);
+  TaylorVecXd this_x = MakeInputTaylorVec(xvec, *(wrapped->variable_list));
+  c->eval(this_x, ty);
+  double result = ty(0).value() * -1;
+
+  const Eigen::VectorXd& upper_bound = c->getUpperBound();
+  if (upper_bound[0] != std::numeric_limits<double>::infinity()) {
+    result -= upper_bound[0];
+  }
+
+  // TODO sam.creasey How should lower bounds be handled?  We don't
+  // have any tests which use them...
```

Comment:
We should write tests, then.  Maybe we could open a Github ticket (issue) for that.